### PR TITLE
Lighter Runtime

### PR DESCRIPTION
**Pull request by refs**
## What?

Replacing the current os thread-based runtime in favor of supervision trees and the option to run the entire set of extensions under a single process. All prior compatibility still exists and extensions can still run individually unsupervised.

## Known issues

- Config parsing using Viper is broken [see issue](https://github.com/spf13/viper/issues/19) and we have to find a way around this to support config via config files
- The entire flag parsing has to be rethought now that we dug into go-micro.
- Reva services are not being stopped on context cancellation.

There are more todo's and regards on the JIRA ticket.

Ticket: https://jira.owncloud.com/browse/OCIS-1715

## TODO
- [ ] add json tags for config files. Currently configuring Reva extensions, values are case sensitive because they get transpiled to golang type names, and these must match. Look into using struct tags.

## Some Considerations:

- [ ] subcommands MUST also set MICRO_LOG_LEVEL to error.
- [ ] normalize flag parsing (and fix hack of renaming ocis top level for the destination side effect)
- [ ] Metrics endpoint in Proxy changed, now we do not panic. What happens if we kill the proxy and run it again? Same metrics?
- [ ] config file parsing with Viper is no longer possible as viper is not thread-safe (https://github.com/spf13/viper/issues/19)
	- [ ] when an extension runs on supervised mode, ocis parses the global config
- [ ] establish on suture a max number of retries before all initialization comes to a halt.
- [ ] Reva runtime currently does not have support for closing a service when a context is canceled??
- [x] reuse the runtime service struct to control suture.
- [x] shutting down a service implies iterating over all the serviceToken for the given service name and terminating them.
- [ ] ~to avoid this getting out of hands, a supervisor would need to be injected on each supervised service.~
- [ ] ~each service would then add its execute func to the supervisor, and return its token (?)~
- [x] this runtime should only care about start / stop services, for that we use serviceTokens.
- [x] ~normalize the use of panics so that suture can restart services that die.~ return achieves the same effect.
- [x] the runtime should ideally run as an rpc service one can do requests, like the good ol' pman, rest in pieces.
- [x] remove default log flagset values.

## More Considerations
- [x] solve the annoying issue with:
    - [x] ~Failed to shutdown server error="context canceled" server=debug service=proxy~
- scaling on a single node works. Example editing the config:
    - create /etc/ocis/proxy.json
    - start ocis using the runtime
    - modify the proxy.json config HTTP and debug ports
    - ocis run proxy
    - there are 2 proxy instances now
- ~deal with the error:~
    - ~panic: send on a closed channel~
- [tusd] logging format is not in sync with the rest of the services
- [x] ~sharing is broken, could it be the same issue with the home storage config we had a few days ago?~ duplicate config downstream.
- [x] ~when running supervised group run should not listen for OS interruptions~
- does web need: //Flags: flagset.RootWithConfig(cfg), ?

## Docs
---

## Configuration via config files

As soon as [#1762](https://github.com/owncloud/ocis/pull/17620) gets merged, oCIS admins will be able to define a global configuration file in order to configure the entire node. Values present in such file correlate to those of [ocis-pkg/config](https://github.com/owncloud/ocis/blob/5594c142ddb6bd4419b04c89927cdab0f1eb0824/ocis-pkg/config/config.go).

Naturally oCIS offers a wide range of ways to be configured, from admin friendly config files to cli flags and cloud / 12factor environment variables, but with so many sources is really easy to get lost in translation. Let's dig into precedence and operation modes.

### Modes of Operation

In order to understand precedence let's define modes of operation first. Depending on how an extension runs this will be _supervised_ or _unsupervised_. Think Erlang supervisors. Whenever the _ocis single binary_ 's `server` command is invoked, all default embedded extensions that will bootstrap a fully functional oCIS instance with its default in-memory configuration. On the other hand, when an extension is ran standalone (picture an external extension built outside the single binary) with the single binary, i.e: `ocis run idp` then there is no supervision tree looking after this extension. In other words the extension is looking only after itself, and is up to it to recover from panics, restart or fork itself in case of unrecoverable states.

Simply put, `ocis server` runs using goroutines and supervision trees, everything else runs as a single, independent process.

#### Supervised
An extension running supervised will receive and modify a configuration in the following order (from more to less relevant):
1. cli flags
2. environment variables
3. config files

Given the nature of oCIS, cli flags can't make it all the way to subcommands, so oCIS does not propagate cli flags down to extensions. To see the list of supported CLI flags for oCIS run `ocis help`.

#### Unsupervised
- TODO

## Example Global Configuration

```yaml
proxy:
  http:
    addr: "0.0.0.0:6200"
accounts:
  http:
    addr: "0.0.0.0:8888"
Storage:
  Reva:
    StorageMetadata:
      Port:
        HTTPAddr: "0.0.0.0:7777"
```

### CODE REVIEW

**Review by butonic — APPROVED** [review]

File: graph-explorer/pkg/command/server.go [hunk 1]
```diff
@@ -216,7 +216,7 @@ func Server(cfg *config.Config) *cli.Command {
 
 					return nil
 				}, func(err error) {
-					close(stop)
+					//close(stop)
```
[the remark this answers is not in the document]
Comment (butonic) [reply graph-explorer/pkg/command/server.go]: @refs ok I got rid of this while fixing graph and graph explorer ...

[the remark this answers is not in the document]
Comment (refs) [reply graph-explorer/pkg/command/server.go]: Ace! Thanks 🙏

**Comment by butonic** [comment]
@refs I think you got around the config parsing using a mutex, right? Tick it off from the 'known issues' list?

**Comment by butonic** [comment]
ok, to shutdown reva we need a context option on the `runtime.RunWithOptions()`...

**Comment by refs** [comment]
> ok, to shutdown reva we need a context option on the `runtime.RunWithOptions()`... but not as part of this PR

thanks for looking into this one :) I would add a follow-up ticket for the fast lane
